File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
        int S1_Profit = 0, B1_Profit = Integer.MIN_VALUE; // Transaction 1 states
    int S2_Profit = 0, B2_Profit = Integer.MIN_VALUE; // Transaction 2 states
        for (int price : prices) {

                    // Transaction 1 updates (must use results from Transaction 0 = 0)
        S1_Profit = Math.max(S1_Profit, B1_Profit + price); // Sell 1 (based on B1)
        B1_Profit = Math.max(B1_Profit, -price); 
        // Transaction 2 updates (must use results from Transaction 1)
        S2_Profit = Math.max(S2_Profit, B2_Profit + price); // Sell 2 (based on B2)
        B2_Profit = Math.max(B2_Profit, S1_Profit - price); // Buy 2 (based on S1)
        
           // Buy 1 (based on 0 profit)
    }            
    return S2_Profit;
    }
}

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        t1buy = float("inf")
        t2buy = float("inf")
        t1profit = 0
        t2profit = 0

        for price in prices:
            # First transaction
            # t1buy: smallest price to buy for the first transaction
            t1buy = min(t1buy, price)
            # t1profit: maximum profit from the first transaction
            # (current price - cost of first buy)
            t1profit = max(t1profit, price - t1buy)
            
            # Second transaction
            # t2buy: smallest price to buy for the second transaction,
            # adjusted by the profit from the first transaction
            # (current price - t1profit already made)
            t2buy = min(t2buy, price - t1profit)
            # t2profit: maximum profit from two transactions
            # (current price - cost of second buy, considering t1profit)
            t2profit = max(t2profit, price - t2buy)
            
        return t2profit